README.md
# goldeneye

## Overview
The `goldeneye` package provides an R-implementation of the `GoldenEye` algorithm described in _Henelius, Puolamäki, Boström, Asker and Papapetrou. 2014. A peek into the black box: exploring classifiers by randomization_, [available here](https://doi.org/10.1007/s10618-014-0368-8). 

The package also implements the `GoldenEye++` algorithm described in the paper _Henelius, Puolamäki, Karlsson, Zhao, Asker, Boström and Papapetrou. 2015. GoldenEye++: A Closer Look into the Black Box_, [available here](https://doi.org/10.1007/978-3-319-17091-6_5). Please note that a bug was found in the `GoldenEye++` variant impacting the groupings reported in the paper; refer to the erratum below and to the [changelog](changelog.md).

Using the `goldeneye` method it is possible to uncover which attributes in a dataset that a classifier uses jointly for making predictions. A *dataset* is of the form

item | A1 | A2 | A3 | A4 | Class
-----|----|----|----|----|-------
1    | 1  | 1  | 1  | 1  |  1
2    | 1  | 0  | 0  | 0  |  1
3    | 0  | 1  | 1  | 1  |  0
4    | 1  | 1  | 1  | 0  |  1
5    | 1  | 0  | 1  | 1  |  1
6    | 1  | 0  | 1  | 0  |  1

where each row of the dataset is a data item of different attrbutes *A1*, ..., *A4* (in this example) and is associated with a *class label*. Given such a dataset and a classifier (e.g., the *randomForest* classifier or a *support vector machine (SVM)*), the `goldeneye` algorithm finds a grouping of attributes of the form *((A1, A2), (A3))*. In this case, the grouping means that attributes *A1* and *A2* are used jointly by the classifier in making predictions, attribute *A3* is not used in conjunction with other attributes, and attribute *A4* is uninformative for the classification task and is not part of the grouping.

In R, the grouping of attributes is done using the `goldeneye` function in the package, see the examples below.


## Citing
To get a BibTex entry in R type `citation("goldeneye")` when the package is installed.


## Installation
The development version of the `goldeneye` package can be installed as follows.

First install the `devtools`-package and load it:
```
install.packages("devtools")
library(devtools)
```

after which the `goldeneye` package cam be installed:
```
install_bitbucket("aheneliu/goldeneye")
```

## Usage
### Example 1
Investigate which attributes the `randomForest` classifier is using in a synthetic dataset.

```R
## Load libraries
library(goldeneye)
library(randomForest)

## Generate synthetic data
set.seed(42)
data <- gen.data.single.xor()

## Find the attribute grouping
set.seed(42)
res <- goldeneye(data = data, classifier = randomForest)

## View the results
> str(res)
List of 8
 $ S                :List of 3
  ..$ : int [1:3] 6 1 2
  ..$ : int [1:2] 6 3
  ..$ : int [1:2] 6 4
 $ S.goodness       : num 0.903
 $ S.pruned         :List of 3
  ..$ : int [1:3] 6 1 2
  ..$ : int [1:2] 6 3
  ..$ : num [1:2] 0 4
 $ S.goodness.pruned: num 0.905
 $ bayes.goodness   : num 0.755
 $ acc.original     : num 0.915
 $ acc.final        : num 0.827
 $ delta            : num 0.0448
```
The grouping is found in `res$S` which in this case corresponds to the grouping `((A1, A2), (A3), (A4))`. The pruned grouping, in which uninformative singleton attributes are removed, is found in `res$S.pruned`. In this case it corresponds to `((A1, A2), (A3))`, i.e., attribute _A4_ is not used by the classifier.

### Example 2
Using classifiers with custom parameters is possible by first creating the model and passing it to the `goldeneye` function.
```R
# Load libraries
library(goldeneye)
library(RWeka)

# Create synthetic data
set.seed(42)
data <- gen.data.single.xor()

# Split data into training and testing
set.seed(42)
train <- testsplit(data)

# Create classifier model: SVM with radial kernel from Weka
set.seed(42)
model <- SMO(Class~., data[train, -(ncol(data))], control = Weka_control(K = "weka.classifiers.functions.supportVector.RBFKernel -G 3"))
data[,"PClass"] <- predict(model, newdata = data)

## Find the attribute grouping
res <- goldeneye(data = data[-train,], model = model)
```

### Example 3
Using a custom goodness function and the randomForest classifier from the randomForest package as in the `GoldenEye++`-algorithm.
```R
# Load libraries
library(goldeneye)
library(randomForest)

## Find the attribute grouping
res <- goldeneye(data = data, classifier = randomForest, goodness.function = class_probability_correlation_randomforest)
```

## License
The `goldeneye` R-package is licensed under the [MIT-license](http://opensource.org/licenses/MIT).

## Erratum
### 2017-12-05
Due to a bug in the code for the `GoldenEye++` algorithm the data was not permuted within-class. Please note that this affects the `GoldenEye++`-groupings reported in the [GoldenEye++ -paper](https://doi.org/10.1007/978-3-319-17091-6_5) which hence are incorrect. Please also refer to the [changelog](CHANGELOG.md).


CHANGELOG.md
# 0.2.1 (2017-12-05)
* Fixed bug in the `GoldenEye++` variant of the algorithm, i.e., when a custom goodness function is used. The data was not permuted within-class as it should be.
* In the `goldeneye`-function, the argument `predict.function` has been removed. The custom prediction using class probabilities instead of class labels is now instead incorporated directly into the custom goodness functions `class_probability_ranking_randomforest` and `class_probability_correlation_randomforest`.
* Renamed the function `class_probability_ranking` to `class_probability_ranking_randomforest`.
* Renamed the function `class_probability_ranking_pearson` to `class_probability_correlation_randomforest`.
* Added README file.
* Added CITATION file.

# 0.2.0 (2017-01-31)
* Fixed bug in the grouping algorithm.

# 0.1.3 (2014-11-19)
* Fixed bug in which the function used to prune singletons always used fidelity instead of the supplied goodness function.
* Fixed wrong correlation type in the function `class_probability_ranking`. The correct correlation is Spearman.

# 0.1.3 (2014-11-07)
* The class column in the input dataset automatically made into a factor.
* Added support for generic goodness functions, to be used instead of fidelity.
* Added support for custom predict functions, to support more advanced usage.

# 0.1.2 (2014-11-05)
* Predicted class automatically added to the data frame if not present.
* Improved check of input arguments.
* Fixed a bug that prevented use of the `svm` classifier from the `e1071` package with `GoldenEye`.
* Updated citation information.

# 0.1.0 (2014-04-25)
* First release.


R/artificial_data.R
#' Basic helper function to generate random data.
#'
#' @param nrows Number of rows in the data 
#' @param ncols Number of columns in the data
#' 
#' @return A binary matrix with random data.
#' 
#' @keywords internal
gen.data.basic <- function(nrows = 1000, ncols = 4) {
    data <- matrix(sample(c(0, 1), nrows * ncols, replace = TRUE), nrow = nrows, ncol = ncols)
    data
}


#' Flip the class label in noiselevel fraction of the data
#'
#' @param classvector A vector giving the class of the data
#' @param noiselevel Fraction of class labels to randomly flip.
#' @return Class vector with randomly noiselevel amount of random noise.
#' 
#' @keywords internal
add.data.noise <- function(classvector, noiselevel) {
    ind <- sample(1:length(classvector), size = floor(noiselevel * length(classvector)), replace = FALSE)
    classvector[ind] <- as.numeric(xor(1, classvector[ind]))
    classvector
}


#' Generate random binary data (Single-XOR)
#'
#' @param nrows Number of rows in the data 
#' @param ncols Number of columns in the data
#' @param noiselevel Fraction of class labels to randomly flip.
#' @param noise Add noise or not (Boolean).
#' @return Data matrix where Class = XOR(1, 2) | 3
#'
#' @export
gen.data.single.xor <- function(nrows = 1000, ncols = 5, noiselevel = 0.1, noise = TRUE) {
    data <- gen.data.basic(nrows = nrows, ncols = ncols)

    ## Link columns
    data[,5] <- xor(data[,1], data[,2]) | data[,3]

    data[,4] <- rep(c(1,0), ceiling(nrows / 2))[1:nrows]
    
    ## Add noise: randomly flip the class bit
    if (noise)
        data[,5] <- add.data.noise(data[,5], noiselevel = noiselevel)

    data           <- data.frame(data)
    colnames(data) <- c(sapply(1:(ncols - 1), function(i) paste("A", i, sep = "")), "Class")
    data$Class     <- factor(data$Class)

    data
}


#' Generate random binary data (Double-XOR)
#'
#' @param nrows Number of rows in the data 
#' @param ncols Number of columns in the data
#' @param noiselevel Fraction of class labels to randomly flip.
#' @param noise Add noise or not (Boolean).
#' @return Data matrix where Class = XOR(1, 2) | XOR(3, 4)
#'
#' @export
gen.data.double.xor <- function(nrows = 1000, ncols = 7, noiselevel = 0.1, noise = TRUE) {
    ncols <- max(7, ncols)

    data <- gen.data.basic(nrows = nrows, ncols = ncols)

    X <- as.numeric(xor(data[,1], data[,2]))
    Y <- as.numeric(xor(data[,3], data[,4]))

    ## Set the class variable
    data[, 7]      <- X | Y

    ## Add noise: randomly flip the class bit
    if (noise)
        data[, 7]      <- add.data.noise(data[,7], noiselevel = noiselevel)

    data           <- data.frame(data)
    colnames(data) <- c(sapply(1:(ncols - 1), function(i) paste("A", i, sep = "")), "Class")
    data$Class     <- factor(data$Class)

    data
}


#' Exact classifier for the Single-XOR data
#'
#' @param formula Model formula
#' @param data The data
#'
#' @export
classifier.single.xor <- function(formula, data) {
    z         <- list()
    z$formula <- formula
    z$data    <- data
    class(z)  <- "classifier.single.xor"
    z
}


#' Exact classifier for the Double-XOR data
#'
#' @param formula Model formula
#' @param data The data
#' 
#' @export
classifier.double.xor <- function(formula, data) {
    z         <- list()
    z$formula <- formula
    z$data    <- data
    class(z)  <- "classifier.double.xor"
    z
}


#' Prediction function for the Single-XOR
#'
#' @param object Object
#' @param newdata The data to predict
#' @param ... Additional arguments
#'
#' @export
predict.classifier.single.xor <- function(object, newdata, ...) {
    as.numeric(xor(newdata[,1], newdata[,2]) | newdata[,3])
}


#' Prediction function for the Double-XOR data
#'
#' @param object Object
#' @param newdata The data to predict
#' @param ... Additional arguments
#'
#' @export
predict.classifier.double.xor <- function(object, newdata, ...) {
    as.numeric(xor(newdata[,1], newdata[,2]) | xor(newdata[,3], newdata[,4]))
}


R/goldeneye_algorithm.R
#' An iterative algorithm to the find the optimal grouping of attributes
#' for a dataset.
#'
#' This function iteratively finds the optimal grouping of attributes for a dataset
#' using a given classifier.
#'
#' @param data The dataset.
#' @param model A model created from the data using a classifier.
#' @param delta The sensitivity parameter. Defaults to 1 / sqrt(nrow(data)).
#' @param class.name The name of the column in the dataset with the predicted classes. Optional. Default is \code{PClass}.
#' @param class.name.2 The name of the column in the dataset with the real classes. Optional. Default is \code{Class}.
#' @param goodness.function The function used to investigate the effect of randomizing attributes in the dataset. Optional. Default is \code{fidelity}.
#' @param return.data Should the inflated dataset used in the grouping process be returned. Default is false.
#'
#' @return A list containing the following items:
#' \describe{
#' \item{S}{The optimal grouping}
#' \item{bayes.goodness}{The goodness (e.g., fidelity) when calculated using a naive Bayes (all-singleton) grouping.}
#' \item{delta}{The value of delta used.}
#' \item{data.inflated}{The dataset, created by sampling with replacement from the testing data, inflated based on the value of delta. Returned if return.data is TRUE.}
#' }
#'
#' @examples
#' data <- gen.data.single.xor()
#' train <- testsplit(data)
#' model <- classifier.single.xor(Class~., data = data[train, -ncol(data)])
#' data[,"PClass"]  <- predict(model, newdata = data)
#' res <- grouping(data = data, model = model)
#'
#' @export
grouping <- function(data, model, delta = (1 / sqrt(nrow(data))), class.name = "PClass", class.name.2 = "Class", goodness.function = fidelity, return.data = FALSE) {
    ## Determine the index of the column class.name
    ## and the indices of the attribute columns
    class.index <- which(names(data) == class.name)
    attrs       <- setdiff(seq.int(ncol(data)), which(names(data) %in% c(class.name, class.name.2)))

    ## Compute Naive Bayes fidelity using an inflated dataset, to ensure precision
    p     <- 0.5
    sigma <- delta / 5
    nr    <- ceiling((p * (1 - p)) / (sigma^2))
    nbfid <- goodness.function(model = model, tree = lapply(attrs, function(i) c(class.index, i)), data = data[sample(seq.int(nrow(data)), nr, replace = TRUE), ], class = class.name)

    ## Inflate the dataset to ensure that the desired variance level can be reached
    p    <- max(0.5, nbfid)
    nr   <- max(1000, ceiling((p * (1 - p)) / (sigma^2)))
    data <- data[sample(seq.int(nrow(data)), nr, replace = TRUE), ]

    ## --------------------------------------------------

    S     <- list()          # accumulated attribute groups
    R     <- as.list(attrs)  # groups to test for
    A     <- list()          # removed attributes
    Delta <- nbfid + delta   # grouping threshold

    ## --------------------------------------------------
    ## group the attributes
    ## --------------------------------------------------

    while ((length(R) > 0) | (length(A) > 0)) {
        if ((length(A) == 0) & (goodness.function(model = model, data = data, class = class.name, tree = maketree(R, S, class.index)) < Delta)) {
            ## Already below Delta before removing any attributes,
            ## so assign remaining attributes to singleton groups
            ##  S <- c(lapply(S, function(i) c(class.index, i)), singletonize(R, class.index))
            S <- c(lapply(S, function(i) c(NULL, i)), singletonize(R, NULL))
            R <- list()
            A <- list()
        } else {
            ## Find the attribute j removal of which from R decreases the fidelity the least
            fids <- sapply(seq.int(length(R)),
                           function(i)
                           goodness.function(model = model,
                                             data  = data,
                                             class = class.name,
                                             tree  = maketree.attr.i(R, c(A, S), i, class.index)))
            idx <- which.max(fids)
            if ((length(R) == 1) | (fids[idx] < Delta)) {
                ## If the fidelity drops below Delta add the group of attributes to the result
                ## and look for the next group of attributes
                S <- c(S, list(unlist(R)))
                R <- A
                A <- list()
            } else {
                ## If the fidelity stays above Delta, then continue by removing the grouping R
                A <- c(A, R[idx])
                R <- setdiff(R, R[idx])
            }
        }
    }

    S <- lapply(S, function(i) c(class.index, i))
    
    ## Return final grouping
    out <- list("S"              = S,
                "bayes.goodness" = nbfid,
                "delta"          = delta)

    if (return.data)
        out <- c(out, list("data.inflated" = data))

    out
}


#' An iterative algorithm to prune singletons.
#'
#' This function iteratively prunes those singletons from a given grouping
#' that do not affect the fidelity more than the given sensitivity.
#'
#' @param data The dataset.
#' @param model A model created from the data using a classifier.
#' @param delta The sensitivity parameter. Defaults to 1 / sqrn(nrow(data)).
#' @param grouping A grouping parametrizing the data.
#' @param class.name The name of the column in the dataset with the predicted classes.
#' @param goodness.function The function used to investigate the effect of randomizing attributes in the dataset. Optional. Default is \code{fidelity}.
#'
#' @return A list containing the optimal grouping, with singletons pruned.
#'
#' @examples
#' data <- gen.data.single.xor()
#' train <- testsplit(data)
#' model <- classifier.single.xor(Class~., data = data[train, -ncol(data)])
#' data[,"PClass"]  <- predict(model, newdata = data)
#' res <- grouping(data = data, model = model)
#' S.pruned <- prune.singletons(data = data, model = model, grouping = res$S)
#'
#' @export
prune.singletons <- function(data, model, delta = (1 / sqrt(nrow(data))), grouping = NULL, class.name = "PClass", goodness.function = fidelity) {
    ## Sanity check
    if (is.null(grouping))
        stop("No grouping provided!")

    ## Set shorthand for the grouping
    S <- grouping

    ## Index of the column with the predicted class
    class.index <- which(names(data) == class.name)

    ## Original fidelity
    Delta <- goodness.function(model = model, data = data, class = class.name, tree = S) - delta

    ## Singleton attributes
    ind.singleton <- which(sapply(S, function(i) length(i)) == 2)

    ## Prune singletons
    while (length(ind.singleton) > 0) {
        ## Find the singleton X removal of which from S decreases the fidelity the least
        fid <- sapply(seq.int(length(ind.singleton)),
                      function(i)
                          goodness.function(model = model,
                                            data  = data,
                                            class = class.name,
                                            tree  = maketree.prune.i(S, ind.singleton[i], class.old = class.index, class.new = 0)))
        idx <- which.max(fid)

        if (fid[idx] >= Delta) {
            ## Prune the singleton and continue
            S <- maketree.prune.i(S, ind.singleton[idx], class.old = class.index, class.new = 0)
            ind.singleton <- setdiff(ind.singleton, ind.singleton[idx])
        } else {
            ## Pruning the singleton would reduce the fidelity too much
            ind.singleton <- c()
        }
    }
    ## Return final grouping
    S
}


#' An iterative algorithm to the find the optimal grouping of attributes
#' for a dataset, also pruning singletons.
#'
#' This function iteratively finds the optimal grouping of attributes for a dataset
#' using a given classifier, after which singletons are pruned.
#'
#' @param data The dataset.
#' @param model A model created from the data using a classifier.
#' @param classifier A classifier (e.g. \code{svm}) to use
#' @param delta The sensitivity parameter. Defaults to 1 / sqrn(nrow(data)).
#' @param real.class.name The name of the column in the dataset with the true classes.
#' @param goodness.function The function used to investigate the effect of randomizing attributes in the dataset. Optional. Default is \code{fidelity}.
#' @param pred.class.name The name of the column in the dataset with the predicted classes. Optional. Default is \code{PClass}. If a column named \code{PClass} is not found in the dataset, it is automatically created.
#' @param return.model Should the model created from a classifier be returned. Default is FALSE.
#' @param return.data Should the data with predictions from the classifier be returned. Default is FALSE.
#'
#' @return A list containing the following items:
#' \describe{
#' \item{S}{The optimal grouping}
#' \item{S.goodness}{The goodness (e.g, fidelity) for the optimal grouping}
#' \item{S.pruned}{The optimal grouping with singletons pruned}
#' \item{S.goodness.pruned}{The goodness (e.g., fidelity) for the optimal grouping with singletons pruned}
#' \item{bayes.goodness}{The  goodness (e.g., fidelity) when calculated using a naive Bayes (all-singleton) grouping.}
#' \item{acc.original}{The original classification accuracy.}
#' \item{acc.final}{The classification accuracy using the final grouping, with singletons pruned.}
#' \item{delta}{The value of delta used.}
#' }
#' In addition, depending on the input arguments, the following fields may also be returned:
#' \describe{
#' \item{model}{The classifier model. Returned if return.model is TRUE.}
#' \item{data}{The testing data separated from the input data. Returned if return.data is TRUE.}
#' \item{data.inflated}{The dataset, created by sampling with replacement from the testing data, inflated based on the value of delta. Returned if return.data is TRUE.}
#' }
#'
#' @examples
#' ## Example 1
#' library(RWeka)
#' set.seed(42)
#' data <- gen.data.single.xor()
#' set.seed(42)
#' goldeneye(data = data, classifier = classifier.single.xor)
#'
#' ## Example 2
#' ## C4.5 Decision tree
#' set.seed(42)
#' goldeneye(data = data, classifier = J48)
#'
#' ## Example 3
#' ## SVM with linear kernel
#' goldeneye(data = data, classifier = SMO)
#'
#' ## Example 4
#' ## Using classifiers with custom parameters is possible by first creating the model
#' ## SVM with radial kernel
#' set.seed(42)
#' train <- testsplit(data)
#' set.seed(42)
#' model <- SMO(Class~., data[train, -(ncol(data))], control =
#'           Weka_control(K = "weka.classifiers.functions.supportVector.RBFKernel -G 3"))
#' data[,"PClass"] <- predict(model, newdata = data)
#' goldeneye(data = data[-train,], model = model)
#'
#' ## Example 5
#' ## Using a custom goodness function and the randomForest classifier from the randomForest package.
#' library(randomForest)
#' res <- goldeneye(data = data, classifier = randomForest, goodness.function = class_probability_correlation_randomforest)
#' 
#' @export
goldeneye <- function(data, model = NULL, classifier = NULL, delta = (1 / sqrt(nrow(data))), real.class.name = "Class", pred.class.name = "PClass", goodness.function = fidelity, return.model = FALSE, return.data = FALSE) {
    if (is.null(model) & is.null(classifier)) {
        stop("No model or classifier provided! Either must be provided.")
    }

    ## Construct a model from the given classifier
    if (is.null(model) & ! is.null(classifier)) {
        cat("Creating classifier model.\n")

        ## Check that the real class name and the predicted class are found.
        ## Predicted class name is created if the default is null.
        if (! real.class.name %in% names(data))
            stop("Real class name not found in the dataset")

        if ((pred.class.name == "PClass") & ! (pred.class.name %in% names(data)))
            data$PClass <- NA
        
        if (! (pred.class.name %in% names(data)))
            stop(paste("Predicted class '", pred.class.name, "' not found in the data.", sep = ""))

        if (! is.factor(data[real.class.name]))
            data[[real.class.name]] <- as.factor(data[[real.class.name]])

        ## if (! is.factor(data[pred.class.name]))
        ##     data[[pred.class.name]] <- factor(data[[pred.class.name]], levels = levels(data[[real.class.name]]))

        ## Index of the column with the predicted class
        real.class.index <- which(names(data) == real.class.name)
        pred.class.index <- which(names(data) == pred.class.name)

        ## Split data into training and testing
        train <- testsplit(data, stratify = real.class.name)

        ## Create model for the data
        model <- eval(parse(text = paste("classifier(", real.class.name, "~., data = data[train, -pred.class.index])", sep = "")))

        ## Add the original accuracy to the dataset
        data[, pred.class.name]  <- predict(model, newdata = data[, -pred.class.index])

        ## Remove the training set from the data
        data <- data[-train, ]

    }

    ## Find the optimal grouping of the dataset
    res        <- grouping(data = data, model = model, class.name = pred.class.name, class.name.2 = real.class.name, delta = delta, goodness.function = goodness.function, return.data = TRUE)
    S.goodness <- goodness.function(model = model, data = res$data.inflated, tree = res$S, class = pred.class.name)

    ## Prune singletons
    S.pruned          <- prune.singletons(data = res$data.inflated, model = model, grouping = res$S, class.name = pred.class.name, goodness.function = goodness.function)
    S.goodness.pruned <- goodness.function(model = model, data = res$data.inflated, tree = S.pruned, class = pred.class.name)

    ## Original accuracy and final accuracy
    acc.original <- sampleaccuracy(model = model, data = res$data.inflated, tree = list(), class = real.class.name)
    acc.final    <- sampleaccuracy(model = model, data = res$data.inflated, tree = S.pruned, class = real.class.name)

    ## Return results
    out <- list("S"                 = res$S,
                "S.goodness"        = S.goodness,
                "S.pruned"          = S.pruned,
                "S.goodness.pruned" = S.goodness.pruned,
                "bayes.goodness"    = res$bayes.goodness,
                "acc.original"      = acc.original,
                "acc.final"         = acc.final,
                "delta"             = res$delta)

    if (return.model)
        out <- c(out, "model" = list(model))

    if (return.data)
        out <- c(out, "data" = list(data), "data.inflated" = list(res$data.inflated))

    out
}
